crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -894,19 +894,26 @@ where
         self.database.history_by_block_hash(block_hash)
     }
 
-    fn state_by_block_hash(&self, block: BlockHash) -> ProviderResult<StateProviderBox> {
-        trace!(target: "providers::blockchain", ?block, "Getting state by block hash");
-        let mut state = self.history_by_block_hash(block);
-
-        // we failed to get the state by hash, from disk, hash block be the pending block
-        if state.is_err() {
-            if let Ok(Some(pending)) = self.pending_state_by_hash(block) {
-                // we found pending block by hash
-                state = Ok(pending)
-            }
+    fn state_by_block_hash(&self, hash: BlockHash) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?hash, "Getting state by block hash");
+        if let Ok(state) = self.history_by_block_hash(hash) {
+            // This could be tracked by a historical block
+            Ok(state)
+        } else if self.canonical_in_memory_state.state_by_hash(hash).is_some() {
+            // ... or this could be tracked by the in memory state
+            let last_persisted_block_number = self.database.last_block_number()?;
+            let latest_historical =
+                self.database.history_by_block_number(last_persisted_block_number)?;
+            let state_provider =
+                self.canonical_in_memory_state.state_provider(hash, latest_historical);
```

Comment:
this is susceptible to a race condition, 

The BlockState has a list of all in memory parents and the on disk anchor can be obtained via:

https://github.com/paradigmxyz/reth/blob/eda161af2c4cb3457eed9d3b81b9b2e5f55db313/crates/chain-state/src/in_memory.rs#L569-L569

so we need to use this here to find the proper disk number

Reply:
ah yeah, the anchor makes sense here